### PR TITLE
Support arbitrary token streams in list attributes.

### PR DESCRIPTION
**Pull request by withoutboats**
With this feature, given an attribute `#[foo(..)]`, the `..` can be an
arbitrary token stream.

### CODE REVIEW

**Comment by withoutboats** [comment]
Added a test which doesn't pass, still figuring out what's wrong here.

**Comment by nikomatsakis** [comment]
@withoutboats how goes -- any progress here?

**Comment by nikomatsakis** [comment]
@withoutboats would you like someone to take a closer look? (I've not looked closely at this PR yet)

**Comment by withoutboats** [comment]
@nikomatsakis Would be good!

I thought adding a fallback to `MetaItem` for cases where it doesn't conform to "nested meta item" syntax would be sufficient (while still supporting the use cases where we evaluate nested meta items in libsyntax today), but its still not parsing correctly and I got lost in the parser trying to figure out where the error was arising from.

**Comment by nikomatsakis** [comment]
cc @petrochenkov -- any thoughts on this?

(I'll try to dig into it later on, ran out of time today.)

**Comment by petrochenkov** [comment]
@nikomatsakis 
>any thoughts on this?

I didn't follow development of token streams and procedural macros 2.0, so I'm not aware of any possible subtleties this change can result in, but in general, if custom attributes are procedural macros, then it seems they should indeed accept arbitrary token trees.

Possible issues:
- Arbitrary token trees in attributes that are *not* procedural macros, attribute validation.
- `MetaItemKind::TokenStream` seems to subsume other variants except for `NameValue` maybe, how are they disambiguated during parsing?

Well, at least it's feature gated, so it's won't be a big problem if something goes unexpectedly wrong.

**Comment by withoutboats** [comment]
@petrochenkov The way I intended the code to work (though it seems like it isn't) is to fallback to TokenStream if it can't parse the attribute as a NestedMetaItem. That way anything that conforms to our existing built in attributes will continue to be parsed as a NestedMetaItem, avoiding rewriting all the code that walks that variant to process built in attributes.

**Comment by withoutboats** [comment]
Also I'd like to stabilize this feature this cycle so that the custom attributes registered by derive can use this syntax.

**Comment by jseyfried** [comment]
Sorry for delay getting to this; I'll investigate test failure this weekend.

**Comment by carols10cents** [comment]
Triage ping -- looks like this is in @nikomatsakis and @jseyfried 's courts to investigate?

**Comment by nikomatsakis** [comment]
OK, @withoutboats, I'm building a copy of this branch locally and will try to investigate.

**Comment by nikomatsakis** [comment]
@withoutboats I took a peek. I believe the problem is that the parser "eagerly commits" to a parse error sometimes. Fixing this will require some mild refactoring. Let me show you what I mean.

As part of parsing a metadata item, we invoke [`parse_seq_to_end`](https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/attr.rs#L279-L281). This invokes [`parse_seq_to_before_end`](https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/parser.rs#L987), which invokes [`parse_seq_to_before_tokens`](https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/parser.rs#L1002).  But something curious happens along the way -- `parse_seq_to_end` returns a `PResult<>`, where the `Err` case is an unemitted error. But if you look closely you will see it is [always returning `Ok`:

https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/parser.rs#L980-L990

`parse_seq_to_before_end`, in contrast, does not return a `Vec`:

https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/parser.rs#L995-L1000

In fact, it invokes `parse_seq_to_before_tokens`, which takes a closure that declares what to do in the case of error. In this case, [it is invoking `|mut e| e.emit()`](https://github.com/withoutboats/rust/blob/cbf5d39cca2e837c7a9880e69e110e714d19c6aa/src/libsyntax/parse/parser.rs#L1002), which will emit the error out. 

So basically this is the problem. We need to refactor `parse_seq_to_before_tokens` to return a `PResult` -- probably what we want to do is to modify the callback so that it can return a `PResult<()>`. However, I don't fully understand the interaction here with parser fallback -- that is, if an error occurs, and we want to keep productively parsing, there may be some interaction here. @nrc may know more on this particular topic.

**Comment by shepmaster** [comment]
Howdy from triage, @withoutboats! It's been over a week since we last heard from you here. Will you have time to work on the test failures soon?

**Comment by shepmaster** [comment]
Thanks for the pr @withoutboats ! Unfortunately, we haven't heard from you in a few weeks, so I'm going to close this to keep things tidy. Please reopen when you have a chance to address the remaining issues!
